crates/wiggle/src/lib.rs

```diff
@@ -659,29 +754,71 @@ impl<T: ?Sized + Pointee> fmt::Debug for GuestPtr<'_, T> {
     }
 }
 
+/// A smart pointer to an immutable slice in guest memory.
+/// Usable as a `&'a [T]` via [`std::ops::Deref`].
+pub struct GuestSlice<'a, T> {
+    ptr: &'a [T],
+    mem: &'a dyn GuestMemory,
+    borrow: BorrowHandle,
+}
+
+impl<'a, T> std::ops::Deref for GuestSlice<'a, T> {
+    type Target = [T];
+    fn deref(&self) -> &Self::Target {
+        self.ptr
+    }
+}
+
+impl<'a, T> Drop for GuestSlice<'a, T> {
+    fn drop(&mut self) {
+        self.mem.unborrow(self.borrow)
+    }
+}
+
 /// A smart pointer to a mutable slice in guest memory.
 /// Usable as a `&'a [T]` via [`std::ops::Deref`] and as a `&'a mut [T]` via
 /// [`std::ops::DerefMut`].
-pub struct GuestSlice<'a, T> {
+pub struct GuestSliceMut<'a, T> {
     ptr: &'a mut [T],
     mem: &'a dyn GuestMemory,
     borrow: BorrowHandle,
 }
 
-impl<'a, T> std::ops::Deref for GuestSlice<'a, T> {
+impl<'a, T> std::ops::Deref for GuestSliceMut<'a, T> {
     type Target = [T];
     fn deref(&self) -> &Self::Target {
         self.ptr
     }
 }
 
-impl<'a, T> std::ops::DerefMut for GuestSlice<'a, T> {
+impl<'a, T> std::ops::DerefMut for GuestSliceMut<'a, T> {
     fn deref_mut(&mut self) -> &mut Self::Target {
         self.ptr
     }
 }
 
-impl<'a, T> Drop for GuestSlice<'a, T> {
+impl<'a, T> Drop for GuestSliceMut<'a, T> {
+    fn drop(&mut self) {
+        self.mem.unborrow(self.borrow)
+    }
+}
+
+/// A smart pointer to an immutable `str` in guest memory.
+/// Usable as a `&'a str` via [`std::ops::Deref`].
+pub struct GuestStr<'a> {
+    ptr: &'a str,
+    mem: &'a dyn GuestMemory,
+    borrow: BorrowHandle,
+}
+
+impl<'a> std::ops::Deref for GuestStr<'a> {
+    type Target = str;
+    fn deref(&self) -> &Self::Target {
+        self.ptr
+    }
+}
+
+impl<'a> Drop for GuestStr<'a> {
     fn drop(&mut self) {
         self.mem.unborrow(self.borrow)
```

Comment:
We technically statically know whether this is a shared or mutable borrow, so could there be two unborrow methods instead of one that checks both maps?